substrate/frame/contracts/src/xcm/adapter.rs

```diff
@@ -0,0 +1,135 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Adapter exposes an implementation of the [`Xcm`] trait using pallet-xcm.
+
+use crate::{
+	xcm::{CallOf, WeightInfo, Xcm},
+	AccountIdOf, Box, Config, RawOrigin,
+};
+use frame_support::{
+	pallet_prelude::{DispatchResultWithPostInfo, IsType},
+	weights::Weight,
+};
+use frame_system::pallet_prelude::BlockNumberFor;
+use pallet_xcm::WeightInfo as XcmWeightInfo;
+use sp_runtime::{DispatchError, DispatchResult};
+use xcm::{v3::MultiLocation, VersionedMultiLocation, VersionedXcm};
+use xcm_executor::traits::{QueryHandler, QueryResponseStatus};
+
+pub type XcmCallOf<T> = <T as pallet_xcm::Config>::RuntimeCall;
+
+/// An implementation of the [`Xcm`] trait using pallet-xcm.
+pub struct XcmAdapter<T: pallet_xcm::Config>(sp_std::marker::PhantomData<T>);
+
+impl<T: pallet_xcm::Config> WeightInfo for XcmAdapter<T> {
+	fn execute() -> Weight {
+		<T as pallet_xcm::Config>::WeightInfo::execute()
+	}
+	fn send() -> Weight {
+		<T as pallet_xcm::Config>::WeightInfo::send()
+	}
+	fn query() -> Weight {
+		<T as pallet_xcm::Config>::WeightInfo::new_query()
+	}
+	fn take_response() -> Weight {
+		<T as pallet_xcm::Config>::WeightInfo::take_response()
+	}
+}
+
+impl<T> XcmAdapter<T>
+where
+	T: Config + pallet_xcm::Config,
+	<T as pallet_xcm::Config>::RuntimeCall: IsType<<T as frame_system::Config>::RuntimeCall>,
+{
+	/// Ensure that the message is executable, by checking that it does not contain any [`Transact`]
+	/// instruction with a call that is not allowed by the CallFilter.
+	fn ensure_executable(
+		message: VersionedXcm<CallOf<T>>,
+	) -> Result<Box<VersionedXcm<XcmCallOf<T>>>, DispatchError> {
+		use frame_support::traits::Contains;
+		use xcm::prelude::{Transact, Xcm};
+
+		let mut message: Xcm<CallOf<T>> =
+			message.try_into().map_err(|_| pallet_xcm::Error::<T>::BadVersion)?;
+
+		message.iter_mut().try_for_each(|inst| -> Result<(), DispatchError> {
+			let Transact { ref mut call, .. } = inst else { return Ok(()) };
+			let call = call.ensure_decoded().map_err(|_| pallet_xcm::Error::<T>::BadVersion)?;
+
+			if !<T as Config>::CallFilter::contains(call) {
+				return Err(frame_system::Error::<T>::CallFiltered.into())
+			}
+
+			Ok(())
+		})?;
+
+		Ok(Box::new(VersionedXcm::from(message.into())))
+	}
+}
+
+impl<T> Xcm<T> for XcmAdapter<T>
```

Comment:
Yeah this should definitely not need to be here. All the functionality you need should already be in traits, and if it's not, we should have no problem moving it to traits in xcm itself.